contracts/protocol/v1/proxies/P1CurrencyConverterProxy.sol

```diff
@@ -0,0 +1,236 @@
+        IERC20(tokenFrom).safeTransferFrom(
+            msg.sender,
+            exchangeWrapper,
+            tokenFromAmount
+        );
+
+        // Convert fromToken to toToken on the ExchangeWrapper.
+        I_ExchangeWrapper exchangeWrapperContract = I_ExchangeWrapper(exchangeWrapper);
+        uint256 tokenToAmount = exchangeWrapperContract.exchange(
+            self,
+            self,
+            tokenTo,
+            tokenFrom,
+            tokenFromAmount,
+            data
+        );
+
+        // Receive toToken from the ExchangeWrapper.
+        IERC20(tokenTo).safeTransferFrom(
+            exchangeWrapper,
+            self,
+            tokenToAmount
+        );
+
+        // Deposit toToken to the Perpetual.
+        perpetualContract.deposit(
+            account,
+            tokenToAmount
+        );
+
+        // Log the result.
+        emit LogConvertedDeposit(
+            account,
+            perpetual,
+            exchangeWrapper,
+            tokenFrom,
+            tokenTo,
+            tokenFromAmount,
+            tokenToAmount
+        );
+
+        return tokenToAmount;
+    }
+
+    /**
+     * @notice Withdraw margin from a Perpetual, then convert the funds to another currency. Funds
+     *  will be withdrawn from the specified account and transfered to the specified destination.
+     * @dev Emits LogConvertedWithdrawal event.
+     *
+     * @param  account          The account to withdraw from.
+     * @param  destination      The address to send the withdrawn funds to.
+     * @param  perpetual        The PerpetualV1 contract to withdraw from to.
+     * @param  exchangeWrapper  The ExchangeWrapper contract to trade with.
+     * @param  tokenTo          The token to convert to.
+     * @param  tokenFromAmount  The amount of `tokenFrom` tokens to withdraw.
+     * @param  data             Trade parameters for the ExchangeWrapper.
+     */
+    function withdraw(
+        address account,
+        address destination,
+        address perpetual,
+        address exchangeWrapper,
+        address tokenTo,
+        uint256 tokenFromAmount,
+        bytes calldata data
+    )
+        external
+        returns (uint256)
+    {
+        I_PerpetualV1 perpetualContract = I_PerpetualV1(perpetual);
+        address tokenFrom = perpetualContract.getTokenContract();
+        address self = address(this);
+
+        // Verify that the sender has permission to withdraw from the account.
+        require(
+            account == msg.sender || perpetualContract.hasAccountPermissions(account, msg.sender),
+            "msg.sender cannot operate the account"
+        );
+
+        // Withdraw fromToken from the Perpetual.
+        perpetualContract.withdraw(
+            account,
+            self,
+            tokenFromAmount
+        );
+
+        // Send fromToken to the ExchangeWrapper.
+        IERC20(tokenFrom).safeTransfer(
+            exchangeWrapper,
+            tokenFromAmount
+        );
```

Comment:
nit: are we able to send the tokens directly to the exchangeWrapper rather than storing them in this contract first?

Reply:
Good idea